### PR TITLE
Purchase : PO -To Refund

### PR DESCRIPTION
**Pull request by shp-odoo**
Description of the issue/feature this PR addresses:
Task:https://www.odoo.com/web?#id=28055&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
Pad:https://pad.odoo.com/p/r.a936a81a26215e34866e0ab77db1234e
Current behavior before PR:
- the "To Refund" box has no impact when returning a PO receipt

Desired behavior after PR is merged:
- If we check the "To Refund" box when returning a receipt,  it should decrease the quantity received on     the PO.
- If we don't check the "To refund" box when returning a receipt, the quantity received on the PO should be unchanged.
- Return with refund should work if only Purchase app is install with stock.
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pedrobaeza** [comment]
@Yajo, please review this

File: addons/stock/models/stock_move.py [hunk 1]
```diff
@@ -151,6 +151,7 @@ def _default_group_id(self):
     restrict_partner_id = fields.Many2one('res.partner', 'Owner ', help="Technical field used to depict a restriction on the ownership of quants to consider when marking this move as 'done'")
     route_ids = fields.Many2many('stock.location.route', 'stock_location_route_move', 'move_id', 'route_id', 'Destination route', help="Preferred route to be followed by the procurement order")
     warehouse_id = fields.Many2one('stock.warehouse', 'Warehouse', help="Technical field depicting the warehouse to consider for the route selection on the next procurement (if any).")
+    to_refund_so = fields.Boolean(string="To Refund", help='Trigger a decrease of the delivered/received quantity in the associated Sale Order/Purchase Order')
```
Comment (Yajo) [inline addons/stock/models/stock_move.py]: This field should get renamed to `to_refund`, given it is used both in SO and PO now.

**Comment by pedrobaeza** [comment]
cc @Tecnativa

File: addons/purchase/models/account_invoice.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def _prepare_invoice_line_from_po_line(self, line):
         else:
             qty = line.qty_received - line.qty_invoiced
         if float_compare(qty, 0.0, precision_rounding=line.product_uom.rounding) <= 0:
-            qty = 0.0
+            qty = qty
```
Comment (Icallhimtest) [inline addons/purchase/models/account_invoice.py]: these 2 lines look very useful.

Somewhere, something seems wrong.

Comment (atp-odoo) [inline addons/purchase/models/account_invoice.py]: For purchase refund case: 
1) when we receieved all products  and we generate and validate/paid invoice, then we return some products with refund.  A new invoice should be generated with negative value for returned products.  

so for this case we have to change qty..

Comment (Icallhimtest) [inline addons/purchase/models/account_invoice.py]: I mean, without looking at anything else than the code. 

if WhatEverConditionIWant : 
 qty = qty

Just makes no sense, just remove both lines, you'll have the same behavior ...

Comment (pedrobaeza) [inline addons/purchase/models/account_invoice.py]: Things said without irony are always better understood and you avoid unneeded bad tempers.

File: addons/purchase/tests/test_create_picking.py [hunk 3]
```diff
@@ -89,3 +89,50 @@ def test_01_check_double_validation(self):
...
+        res = return_wiz.create_returns()
+        return_pick = self.env['stock.picking'].browse(res['res_id'])
+
+        # Validate picking
+        return_pick.force_assign()
+        return_pick.pack_operation_product_ids.write({'qty_done': 2})
+        return_pick.do_new_transfer()
+
+        # Check delivered quantity
+        self.assertEqual(self.po.order_line[0].qty_received, 3.0, 'Purchase: delivered quantity should be 3.0 instead of "%s" after picking return' % self.po.order_line[0].qty_received)
```
Comment (Icallhimtest) [inline addons/purchase/tests/test_create_picking.py]: It's nice to see tests added with new functionalities :-) 

(though it doesn't look like it tests the created invoices :p )

File: addons/stock/models/stock_move.py [hunk 4]
```diff
@@ -151,6 +151,7 @@ def _default_group_id(self):
     restrict_partner_id = fields.Many2one('res.partner', 'Owner ', help="Technical field used to depict a restriction on the ownership of quants to consider when marking this move as 'done'")
     route_ids = fields.Many2many('stock.location.route', 'stock_location_route_move', 'move_id', 'route_id', 'Destination route', help="Preferred route to be followed by the procurement order")
     warehouse_id = fields.Many2one('stock.warehouse', 'Warehouse', help="Technical field depicting the warehouse to consider for the route selection on the next procurement (if any).")
+    to_refund = fields.Boolean(string="To Refund", help='Trigger a decrease of the delivered/received quantity in the associated Sale Order/Purchase Order')
```
Comment (Icallhimtest) [inline addons/stock/models/stock_move.py]: So much for odoo's modularity, but I guess it's better than defining it twice in sale and purchase.

File: addons/purchase/models/stock.py [hunk 5]
```diff
@@ -60,7 +60,7 @@ def get_price_unit(self):
     def copy(self, default=None):
         self.ensure_one()
         default = default or {}
-        if not default.get('split_from'):
+        if not default.get('split_from') and not default.get('origin_returned_move_id'):
```
Comment (Icallhimtest) [inline addons/purchase/models/stock.py]: Could you explain why we do this ? I'm not familiar with the use case of copying stock moves.

Could be nice to also update the comment that follows.

File: addons/purchase/models/account_invoice.py (hunk 2 above)
Comment (Icallhimtest) [inline addons/purchase/models/account_invoice.py]: Ok, so this actually needs to be 
```python
if float_compare(qty, 0.0, precision_rounding=line.product_uom.rounding) == 0:
     qty = 0.0
```
I don't know why the rounding was needed but apparently it was (otherwise we'd just have `if qty < 0:`).

Apologies if I offended anyone with my first comment. It wasn't my intention.

File: addons/purchase/models/stock.py (hunk 5 above)
Comment (atp-odoo) [inline addons/purchase/models/stock.py]: when do partial receive or refund , and create backorder, at that time
 copy method is called and to link that refund move we have to keep
 purchase line id instead of set it false..

File: addons/purchase/models/purchase.py [hunk 6]
```diff
@@ -536,10 +536,10 @@ def _compute_qty_received(self):
             total = 0.0
             for move in line.move_ids:
                 if move.state == 'done':
-                    if move.product_uom != line.product_uom:
+                    if move.location_dest_id.usage == "internal":
                         total += move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
-                    else:
-                        total += move.product_uom_qty
+                    elif move.location_dest_id.usage == "supplier" and move.to_refund:
+                        total -= move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
```
Comment (Icallhimtest) [inline addons/purchase/models/purchase.py]: I'd actually do it this way:
```python
if move.location_dest_id.usage == "supplier" and move.to_refund:
     total -= move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
else:
     total += move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
```
This way if there's another case of usage (could be from a custom module), it works like before.

Comment (atp-odoo) [inline addons/purchase/models/purchase.py]: no.. as I already explained that condition will increase qty also when we return product without refund and that is wrong.. so i put internal location condition..

Comment (Icallhimtest) [inline addons/purchase/models/purchase.py]: Right I forgot, then

```python
if move.location_dest_id.usage == "supplier":
    if move.to_refund:
         total -= move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
else:
    total += move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
```
or shorter if you prefer

```python
if move.location_dest_id.usage == "supplier":
    total -= move.to_refund and move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
else:
    total += move.product_uom._compute_quantity(move.product_uom_qty, line.product_uom)
```

File: addons/purchase/tests/test_purchase_order.py [hunk 7]
```diff
@@ -16,22 +16,11 @@ def setUp(self):
...
         self.product_id_1 = self.env.ref('product.product_product_8')
         self.product_id_2 = self.env.ref('product.product_product_11')
 
-        # Ensure product_id_2 doesn't have res_partner_1 as supplier
-        if self.partner_id in self.product_id_2.seller_ids.mapped('name'):
-            id_to_remove = self.product_id_2.seller_ids.filtered(lambda r: r.name == self.partner_id).ids[0] if self.product_id_2.seller_ids.filtered(lambda r: r.name == self.partner_id) else False
-            if id_to_remove:
-                self.product_id_2.write({
-                    'seller_ids': [(2, id_to_remove, False)],
-                })
```
Comment (jco-odoo) [inline addons/purchase/tests/test_purchase_order.py]: Why did you need to remove this?

**Comment by jco-odoo** [comment]
A quick review tells me it is ok.

**Comment by pedrobaeza** [comment]
Any update on this?

**Comment by jbeficent** [comment]
The scope of the changes should also include the possibility to create an invoice refund from the PO, when you have already received a bill from the supplier.
 
Also, after triggering a return to refund, the if the purchase was in status 'Invoiced', it should return to 'Waiting Invoices' because it fact you expect the supplier to refund to you.

See https://github.com/OCA/account-invoicing/pull/230

**Comment by jco-odoo** [comment]
And it does not do that already with this pull request?

**Comment by jbeficent** [comment]
@jco-odoo I do not see an option to create refund invoices from a PO, neither in master nor in this PR. Also, I do not see any code that changes the behaviour of invoice_state here

**Comment by jco-odoo** [comment]
@jbeficent There is no invoice_state anymore and you don't create vendor bills from the po as with so, as the way you handle them is different.  

The trouble is it is creating negative invoices by default. These should be put in refunds.  This PR is not ok.

**Comment by jbeficent** [comment]
Hi @jco-odoo I see the invoice state still in the PO, in master (https://github.com/odoo/odoo/blob/master/addons/purchase/models/purchase.py#L136). 

I know that you do not create the vendor bills from the PO as you do from SO. 
As for creating negative invoices by default, I agree that you should create refunds instead.

All this is fixed here: OCA/account-invoicing#230

File: addons/purchase/models/purchase.py (hunk 6 above)
Comment (jco-odoo) [inline addons/purchase/models/purchase.py]: The case with supplier is better, less conditions. Or use the location_id on the purchase to see if it is going out or not?
